### ISSUE TITLE
ENH: Ensure F2PY generates Python-C stable API code

### ISSUE THREAD

**Issue by HaoZeke** [numpy/numpy#21300]
### Proposed new feature or change:

The [C API Stability](https://docs.python.org/3/c-api/stable.html#stable) document defines a subset of functions which are (to some extent) compatible across python versions. Thankfully most of the API functions used in `fortranobject.{c,h}` are already in the stable API.

It would be good to ensure this strictly though. A solution would be effectively to ensure that all C extensions compile with `Py_LIMITED_API` set to the same lowest Python version supported by NumPy itself. [1]

In practice, there is a slight performance hit from not using functions outside the stable API, but more importantly, the reference counting is not always the same between these functions (e.g. The sequence protocol defines an "unsafe" `Fast_GET_ITEM` which borrows a reference instead of `GetItem` which returns a new reference). This means that some care may need to be taken to ensure equivalence.

[1] Also to make sure that none of the "unstable" API functions are used

**Comment by m-newhauser** [comment]
Does this still need working on?

**Comment by HaoZeke** [comment]
> Does this still need working on?

Yes but it's tricky since f2py tests are still less than comprehensive and also at the same time very low priority... I'd not recommend picking this up in the near future.
